rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,509 @@
+
+
+## Sharing state between workers
+
+This is not identified as necessary at the moment, and IPC broadcast should be sufficient, hopefully. 
+
+If we do need shared state, we will probably have to use syscall libraries to share buffers such as [mmap-io](https://www.npmjs.com/package/mmap-io), 
+and expose a higher level API for that from the `clustering` service.
+
+# Technical impact on the codebase
+
+This section is meant to be an exhaustive (even if realistically not at the moment) inventory  of the changes that would
+be required to support clustering mode.
+
+## Technical impact on Core
+
+### Handling multi-process logs
+
+This is an example of log output in a 2 workers cluster, coming from the POC
+
+```
+[2021-03-02T10:23:41.834+01:00][INFO ][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.840+01:00][INFO ][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.900+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
+[2021-03-02T10:23:41.903+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
+```
+
+The workers logs are interleaved, and, most importantly, there is no way to see which process the log is coming from. 
+We will need to address that.
+
+Our options are
+
+- Have a distinct logging configuration for each worker
+
+We could do that by automatically adding a suffix depending on the process to the appropriate appenders configuration. 
+Note that this doesn’t solve the problem for the console appender, which is probably a no-go.
+
+- Add the process info to the log pattern and output it with the log message
+
+We could add the process name information to the log messages, and add a new conversion to be able to display it with the pattern layout, such as %process for example.
+
+the default pattern could evolve to (ideally, only when clustering is enabled)
+
+```
+[%date][%level][%process][%logger] %message
+```
+
+The logging output would then look like
+
+```
+[2021-03-02T10:23:41.834+01:00][INFO ][worker-1][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.840+01:00][INFO ][worker-2][plugins-service] Plugin initialization disabled.
+```
+
+This will require some changes in the logging system implementation, as currently the BaseLogger has no logic to 
+enhance the log record with context information before forwarding it to its appenders, but this still
+seems to be the best solution.
+
+### The rolling-file appender
+
+The rolling process of the `rolling-file` appender is going to be problematic in clustered mode, as it will cause 
+concurrency issues during the rolling. We need to find a way to have this rolling process clustered-proof.
+
+Identified options are:
+
+- have the rolling file appenders coordinate themselves when rolling
```

Comment:
nit: can use you numbering for these options so that it's easier to see what you're referring to below